### PR TITLE
Adding multiplayer/networking

### LINKED ISSUE #3: networking

### LINKED ISSUE #28: actually get mohammed uddin to contribute

### PR DESCRIPTION
**Pull request by jesings**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -6,5 +6,8 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-rand = "0.8.0"
-sdl2 = "0.35.0"
+rand = "0.8"
+sdl2 = "0.35"
+serde = { version = "1.0", features = ["derive"] }
+bincode = "1.3"
+bus = "2.2"
```
Comment (anotherLostKitten) [inline Cargo.toml:13]: more like sus

File: Cargo.toml [hunk 2]
```diff
@@ -6,5 +6,8 @@ edition = "2018"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-rand = "0.8.0"
-sdl2 = "0.35.0"
+rand = "0.8"
```
Comment (anotherLostKitten) [inline Cargo.toml:9]: ayn rand reference

File: src/class/class.rs [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
 
-#[derive(Debug)]
+#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
 pub enum Class {
     VoidOrifice(VoidOrifice)
 }
 
 impl Class {
-    pub fn mov(&mut self, pid: usize, dir: (isize, isize), now: Duration) -> impl Fn(Arc<Gamedata>) -> () {
+    pub fn mov(&mut self, pid: usize, dir: (isize, isize), now: Duration) -> impl Fn(Arc<Gamedata>, &mpsc::Sender<PktPayload>) -> () {
```
Comment (anotherLostKitten) [inline src/class/class.rs]: minepaft sunger cames sender pkt reference

File: src/client_gameloop.rs [hunk 4]
```diff
@@ -18,22 +18,43 @@ fn init_game() -> gamestate::ClientGamestate {
     let canvas = window.into_canvas().build().unwrap();
 
     let event_pump = sdl_context.event_pump().unwrap();
+    let mut upstream = clinet::initialize_client("127.0.0.1:9495".to_string());
```
Comment (anotherLostKitten) [inline src/client_gameloop.rs]: have port as module constant

**Review by anotherLostKitten — APPROVED** [review]
can you also update the readme with info on launching to run either client or server

File: src/main.rs [hunk 5]
```diff
@@ -12,19 +12,29 @@ mod class;
...
+            client_gameloop::gameloop()
+        });
+        if hasserver {
+            server_gameloop::gameloop();
+        }
+        client.join().unwrap().unwrap();
+    } else if hasserver {
         server_gameloop::gameloop();
+    } else {
+        eprintln!("Wtfrick is the point of you running am2 anyway???");
```
Comment (anotherLostKitten) [inline src/main.rs:39]: to see the funny error messages

**Review by anotherLostKitten — APPROVED** [review]

**Review by muddint — APPROVED** [review]
my name is mohammed uddin, and i approve this pr

**Pull request merged**